src/language-js/printer-estree.js

```diff
@@ -490,29 +490,30 @@ function printPathNoParens(path, options, print, args) {
         n.expression.type === "ObjectExpression"
       );
 
-      const typeAnnotationGroup = group(
+      const castGroup = group(
         concat([
+          "<",
           indent(concat([softline, path.call(print, "typeAnnotation")])),
-          softline
+          softline,
+          ">"
         ])
       );
 
+      const exprContents = concat([
+        ifBreak("("),
+        indent(concat([softline, path.call(print, "expression")])),
+        softline,
+        ifBreak(")")
+      ]);
+
       if (shouldBreakAfterCast) {
-        return group(
-          concat([
-            "<",
-            typeAnnotationGroup,
-            ">",
-            ifBreak("("),
-            indent(concat([softline, path.call(print, "expression")])),
```

Comment:
I *think* you could just wrap this in a group. The thing here is, when the `typeAnnotationGroup` breaks, it will also break the parent group, which will then make the behavior you are seeing.

But if you wrap this part here in a group, the fact that the parent is "broken" won't affect a child. It will try to print in a single line before breaking.

I'll try my suggestion in a bit and report back :-)